src/otaclient/ota_core/_update_libs.py

```diff
@@ -0,0 +1,436 @@
+                    session_id=self.session_id,
+                )
+
+                verified_base_db = find_saved_fstable(self._image_meta_dir_on_active)
+                if verified_base_db:
+                    logger.info("use rebuild mode with base file table assist ...")
+                    RebuildDeltaWithBaseFileTable(**_rebuild_mode_params).process_slot(
+                        str(verified_base_db)
+                    )
+                else:
+                    logger.info("use rebuild mode with full scanning ...")
+                    RebuildDeltaGenFullDiskScan(**_rebuild_mode_params).process_slot()
+
+            return all_resource_digests
+        except Exception as e:
+            _err_msg = f"failed to generate delta: {e!r}"
+            logger.exception(_err_msg)
+            raise ota_errors.UpdateDeltaGenerationFailed(
+                _err_msg, module=__name__
+            ) from e
+
+
+def download_resources_handler(
+    downloader: Generator[Future[DownloadResult]],
+    *,
+    metrics: OTAMetricsData,
+    status_report_queue: Queue[StatusReport],
+    session_id: str,
+) -> None:
+    _next_commit_before, _report_batch_cnt = 0, 0
+    _merged_payload = UpdateProgressReport(
+        operation=UpdateProgressReport.Type.DOWNLOAD_REMOTE_COPY
+    )
+    for _done_count, _fut in enumerate(downloader, start=1):
+        _now = time.time()
+        if download_exception_handler(_fut):
+            _download_res = _fut.result()
+
+            _merged_payload.processed_file_num += 1
+            _merged_payload.processed_file_size += _download_res.download_size
+            _merged_payload.errors += _download_res.retry_count
+            _merged_payload.downloaded_bytes += _download_res.traffic_on_wire
+        else:
+            _merged_payload.errors += 1
+
+        if (
+            _this_batch := _done_count // DOWNLOAD_STATS_REPORT_BATCH
+        ) > _report_batch_cnt or _now > _next_commit_before:
+            _next_commit_before = _now + DOWNLOAD_REPORT_INTERVAL
+            _report_batch_cnt = _this_batch
+
+            status_report_queue.put_nowait(
+                StatusReport(
+                    payload=_merged_payload,
+                    session_id=session_id,
+                )
+            )
+
+            metrics.downloaded_bytes += _merged_payload.downloaded_bytes
+            metrics.downloaded_errors += _merged_payload.errors
+            _merged_payload = UpdateProgressReport(
+                operation=UpdateProgressReport.Type.DOWNLOAD_REMOTE_COPY
+            )
+
+    # for left-over items that cannot fill up the batch
+    status_report_queue.put_nowait(
+        StatusReport(
+            payload=_merged_payload,
+            session_id=session_id,
+        )
+    )
+
+
+@contextlib.contextmanager
+def metadata_download_err_handler():
+    """A context manager that parses the exception during metadata files downloading."""
+    try:
+        yield
```

Comment:
A common helper for handling errors during OTA image metadata parsing(non OTA image spec version specific).